docs/design/sql/04-mulitthreaded-execution.md

```diff
@@ -0,0 +1,215 @@
+IMDG and Hazelcast Jet. In section 2 we analyze why the existing infrastructure is inappropriate for query execution and then
+describe the design of the Hazelcast Mustang execution environment. In section 3 we discuss possible alternative approaches
+that were rejected.
+
+## 1 Existing Infrastructure
+
+Hazelcast IMDG uses staged event-driven architecture (SEDA) for message processing. During execution, a message passes through
+several thread pools (stages), each optimized for a specific type of workload. We now describe stages that exist in Hazelcast.
+
+### 1.1 IO Pool
+
+Hazelcast uses a dedicated thread pool for message send and receive, which will be referred to as **IO Pool** in this paper.
+Each thread from the IO pool maintains a subset of connections to remote members. Consider that we have a sender member (S)
+and a receiver member (R). The typical execution flow is organized as follows:
+1. The message is added to the queue of a single IO thread, and the thread is notified
+1. The sender IO thread wakes up and sends the message over the network
+1. A receiver IO thread is notified by the operating system on receive
+1. The receiver IO thread wakes up, determines the next execution stage, adds the message to the stage's queue and notifies the
+stage
+1. The next execution stage processes the message
+
+*Snippet 1: Message execution flow*
+```
+Stage(S)                 IO(S)        IO(R)                Stage(R)
+   |----enqueue/notify->--|            |                      |
+   |                      |----send->--|                      |
+   |                      |            |----enqueue/notify->--|
+```
+
+We now discuss the organization of different execution stages.
+
+### 1.2 Partition Pool
+
+A message may have a logical **partition**, which is a positive integer number. Messages with defined partition are routed to
+a special thread pool, which we refer to as **partition pool**. The pool has several threads. Every thread has a dedicated task
+queue. Partition of the message is used to determine the exact thread which will process the message:
+`threadIndex = partition % threadCount`.
+
+The partition pool has the following advantages:
+1. Only one thread processes messages with the given partition so that processing logic may use less synchronization
+1. Dedicated thread queues reduce contention on enqueue/deque operations
+
+The partition pool has the following disadvantage:
+1. There is no balancing between threads: a single long-running task may delay other tasks from the same partition
+indefinitely; likewise, an imbalance between partitions may cause resource underutilization
+
+The partition pool is thus most suitable for small tasks that operate on independent physical resources, and that are
+distributed equally between logical partitions. An example is `IMap` operations, which update separate physical partitions,
```

Comment:
Probably "..., which operate on separate ..." is a better wording, "update" sounds strange when applied to `GET`.

Reply:
Rephrased.